Desktop/data/undostack.h

```diff
@@ -58,35 +58,44 @@ class SetWorkspacePropertyCommand: public UndoModelCommand
 							_oldValue;
 };
 
+class UndoModelCommandLabelChange: public UndoModelCommand
+{
+public:
+	UndoModelCommandLabelChange(QAbstractItemModel *model);
+	
+	void undo()					override;
+	void redo()					override;
+	
+	
+protected:
+	ColumnModel*			_columnModel = nullptr;
+	int						_colId		= -1;
+	Json::Value				_oldLabels	= Json::nullValue;
+};
+
 
-class SetLabelCommand: public UndoModelCommand
+class SetLabelCommand: public UndoModelCommandLabelChange
 {
 public:
 	SetLabelCommand(QAbstractItemModel *model, int labelIndex, QString newLabel);
 	
-	void undo()					override;
 	void redo()					override;
 	
 private:
-	ColumnModel*			_columnModel = nullptr;
-	int						_colId		= -1,
-							_labelIndex = -1;
+	int						_labelIndex = -1;
 	QString					_newLabel,
 							_oldLabel;
 };
 
-class SetLabelOriginalValueCommand: public UndoModelCommand
+class SetLabelOriginalValueCommand: public UndoModelCommandLabelChange
```

Comment:
I think you forgot to also undo value changes in the data. The undo method must be also overriden.
If you change the value of a label, and undo it, it does not undo the values. The old value/label still exists. For example: In contBinom, change Value 1 to 10, Undo: the 10 are still in the data. Change column, and come back to contBinom: you see still the label 10/10.